r2d7/roller.py

```diff
@@ -0,0 +1,252 @@
+        self.dice = [dieFactory[self.die_type]() for i in range(self.num_dice)]
+
+        try:
+            self.parse_mod_boolean('focus', message)
+            self.parse_mod_boolean('lock', message)
+            self.parse_mod_numeric('evade', message)
+            self.parse_mod_numeric('reinforce', message)
+            self.parse_mod_numeric('calculate', message)
+            self.parse_mod_numeric('force', message)
+            self.parse_mod_numeric('reroll', message)
+
+        except Exception as err:
+            logger.debug('roll parsing error: %s \n message: %s' % (str(err), message))
+            raise RollSyntaxError('Incorrect dice roll syntax. Someone call a judge.')
+
+        if self.reroll > 3:
+            logger.debug('Too many rerolls requested (max: 3, requested: %d)' % (self.reroll))
+            raise RollSyntaxError('Sorry, the calculator only allows up to 3 rerolls')
+
+        self.modify_dice()
+
+    def parse_mod_boolean(self, attr_name, message):
+        value = ModdedRoll.pattern_mod[attr_name].search(message) is not None
+        object.__setattr__(self, attr_name, value)
+
+    def parse_mod_numeric(self, attr_name, message):
+        match_mod = ModdedRoll.pattern_mod[attr_name].search(message)
+        if match_mod:
+            match_count = ModdedRoll.pattern_count.search(match_mod.group(1))
+            value = int(match_count.group(1)) if match_count is not None else 1
+        else:
+            value = 0
+        object.__setattr__(self, attr_name, value)
+
+    def modify_dice(self):
+        if len(self.dice) == 0:
+            return
+
+        #reroll first!
+        if self.die_type == DieType.attack and self.lock:
+            [d.reroll() for d in self.dice]
+        elif self.reroll > 0:
+            rerolls = self.reroll
+            for d in self.dice:
+                rerolls = rerolls - 1 if d.reroll() else rerolls
+                if rerolls == 0:
+                    break
+
+        # after reroll, apply other mods
+        # this is a 1-off calculation so we can spend force greedily
+        calculates = self.calculate
+        force = self.force
+        evades = self.evade
+        for d in self.dice:
+            focussed = False
+            if self.focus:
+                focussed = d.focus()
+            elif calculates > 0:
+                focussed = d.focus()
+                calculates = calculates - 1 if focussed else calculates
+            elif force > 0:
+                focussed = d.focus()
+                force = force - 1 if focussed else force
+            if not focussed and evades > 0 and d.die_type == DieType.defense:
+                evades = evades - 1 if d.evade() else evades
+
+    def roll_safe(self):
+        return len(self.dice) <= CalculatorForm.max_dice and self.reinforce <= CalculatorForm.max_reinforce
+
+    def calculator_form(self):
+        if self.die_type == DieType.attack:
+            form = AttackForm(dice = len(self.dice),
```

Comment:
I suggest putting a newline after `(` so all the kwargs are indented the same amount.